### ISSUE TITLE
Make it possible to declare ASYNCIFY_IMPORTS via a JS library annotation?

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#15746]
I have an asynchronous JS library function:

C:
```c
void navigator_gpu_request_adapter_async(WGpuRequestAdapterCallback adapterCallback);
```
JS:
```js
  navigator_gpu_request_adapter_async: function(adapterCallback) {
    navigator['gpu']['requestAdapter']().then(adapter => {
      {{{ makeDynCall('vi', 'adapterCallback') }}}(wgpuStore(adapter));
    });
  },
```
To make it possible to synchronously initialize a WebGPU adapter, I want to use ASYNCIFY, so I create a new variant of this function, `navigator_gpu_request_adapter_sync` that reads

C:
```c
WGpuAdapter navigator_gpu_request_adapter_sync(void);
```
JS:
```js
  navigator_gpu_request_adapter_sync: function() {
    return Asyncify.handleAsync(() => {
      return navigator['gpu']['requestAdapter']().then(adapter => {
        return wgpuStore(adapter);
      })
    });
  },
```
and build with `-sASYNCIFY=1` linker flag.

However, at runtime I get the error:

```
Uncaught Error: import navigator_gpu_request_adapter_sync was not in ASYNCIFY_IMPORTS, but changed the state
```

Reading the documentation, the currently intended way to fix this would be to add the linker flag

```
-sASYNCIFY_IMPORTS=['navigator_gpu_request_adapter_sync']
```

However I would not like to do that, because of the two problems:
1) it leaks abstractions to the user. Any user of my lib_webgpu.js library will need to know/remember which functions need to be annotated with `-sASYNCIFY_IMPORTS` when they build their program. I.e. one more thing for them to remember to do.
2) if the users have their own asyncified functions, concatenating/merging lib_webpu.js asyncify imports and their own asyncify imports is not currently possible, i.e. I cannot document them to blindly pass `-sASYNCIFY_IMPORTS=['a','b','c']` to their link line, since if they have their own `ASYNCIFY_IMPORTS` as well, either they will override my imports (hence having no effect), or my documented line will replace their imports (breaking their asyncified functions).

Instead, I would like to declare the function to be asynchronous in my JS file

```js
  navigator_gpu_request_adapter_sync__asyncify: true, // Declare this function to require asyncify
  navigator_gpu_request_adapter_sync: function() {
    return Asyncify.handleAsync(() => {
      return navigator['gpu']['requestAdapter']().then(adapter => {
        return wgpuStore(adapter);
      })
    });
  },
```

This way both problems 1) and 2) would be fixed, without asking the people to need to know about the internals of my WebGPU JS library and have them manually maintain these types of lists when using the library.

Would this be feasible to implement? (or does it come down to the same JS library parsing order issue that deps_info.py has?)

**Comment by sbc100** [comment]
Sounds like a good idea to me.  @RReverser has also asked for source annotations to make native functions in needed asyncify.  

I think we are fine with ordering because the JS library is compiled during `phase_emscript` which happens before `phase_binaryen` (which run asyncify and other binaryen passes)

**Comment by RReverser** [comment]
Note that similar usecase is already solved via `EM_ASYNC_JS` macro I've added recently - it allows to declare JS functions that will be automatically added to Asyncify list, plus takes care of the wrapping. But yeah, it's not something available in Emscripten's JS libraries at the moment.

**Comment by sbc100** [comment]
Oh interesting.. it looks like one could already piggyback on what you did and just start give the function a name that starts with with `__asyncjs__`:  https://github.com/emscripten-core/emscripten/blob/69409a6364bd0897337b1aeba32e8609fcda7e11/src/library_async.js#L79-L80 

But using the `__asyncify` tag in the library code sounds good.

**Comment by juj** [comment]
```js
var isAsyncifyImport = ASYNCIFY_IMPORTS.indexOf(x) >= 0 || 
                        x.startsWith('__asyncjs__'); 
```
That raises two thoughts:

1) Doesn't that `x.startsWith('__asyncjs__')` break under Closure minification? Closure should be instructed never to minify any symbol that starts with `__asyncjs__`? (not sure if that's possible for Closure)

2) Is there a need to do this `isAsyncifyImport` check in release/production mode builds? Could that check be put behind ASSERTIONS?

**Comment by sbc100** [comment]
Oh yes, that code I linked to in `emscripten/src/library_async.js` is already inside of `#ifdef ASSERTIONS` so doesn't effect release builds.  

The actual way this is handled is here:
https://github.com/emscripten-core/emscripten/blob/cb545d95b18df50d35ffe5c85e72c709f1d6df1c/emcc.py#L97-L105

**Comment by RReverser** [comment]
Regardless of the resolution here, I would probably encourage looking into Embind's Asyncify integration added earlier as an alternative.

E.g. the code above

```js
navigator_gpu_request_adapter_sync: function() {
    return Asyncify.handleAsync(() => {
      return navigator['gpu']['requestAdapter']().then(adapter => {
        return wgpuStore(adapter);
      })
    });
  },
```

Could be rewritten to retrieve adapter as `val` instance like this:

```cpp
val adapter = val::global("navigator")["gpu"].call<val>("requestAdapter").await();
```

After this `adapter` is already a synchronous handle that you can invoke any other APIs on.

**Comment by sbc100** [comment]
So far we don't have any emscripten-provided libraries that depend on embind.. so have webgpu depend on embind would be departure from that norm.

I'm not sure it makes sense to build our low level system libraries on embind since IIRC it can be somewhat heavy-weight.    Perhaps we could land the version based on the `__asyncify` option then propose switch to embind as a separate issue where we can assess the cost/benefit.

**Comment by RReverser** [comment]
Hmm is that Emscripten-provided library? The only reference to `wgpuStore` I found is coming from this repo, which seems external? https://github.com/juj/wasm_webgpu

**Comment by sbc100** [comment]
Ah I see, yes.   Well I think my point still stands, maybe some folks want to use asyncify to its fullest without also using embind?

**Comment by RReverser** [comment]
Oh yeah, could be. The reason I suggested it in this case is because what `wgpuStore` does (https://github.com/juj/wasm_webgpu/blob/bba80ae7b0dc389f5b3bc05e6816af48781169d8/lib/lib_webgpu.js#L36-L63) is very close to replicating `val`'s behaviour, and it seemed like a nice opportunity to kill two birds with one stone.

But then again, given the amount of code, I can already see it's infeasible endeavour.

**Comment by RReverser** [comment]
FWIW (again, doesn't solve the library use-case, just for the record) yet another alternative would be to use the mentioned `EM_ASYNC_JS` from C++.

Before:
```js
navigator_gpu_request_adapter_sync: function() {
    return Asyncify.handleAsync(() => {
      return navigator['gpu']['requestAdapter']().then(adapter => {
        return wgpuStore(adapter);
      })
    });
  },
```

After:
```cpp
EM_ASYNC_JS(WGpuAdapter, navigator_gpu_request_adapter_sync, (), {
  return wgpuStore(await navigator['gpu']['requestAdapter']());
});
```

**Comment by juj** [comment]
Thanks @RReverser for the suggestion. I am well familiar with embind, I was its maintainer for a while, and also commercially contracted in the past to develop embind back in 2013.

However for this purpose/project, I am not looking to build on top of embind as a dependency, but rather use the direct asyncify API.

> Oh yes, that code I linked to in `emscripten/src/library_async.js` is already inside of `#ifdef ASSERTIONS` so doesn't effect release builds.

Thanks - that reveals I did not much look at the code, but just briefly glanced a few lines around the call site :)

**Comment by RReverser** [comment]
> Thanks @RReverser for the suggestion. I am well familiar with embind, I was its maintainer for a while, and also commercially contracted in the past to develop embind back in 2013.

I'm aware :) But you might've not seen the Embind + Asyncify integrations or new Asyncify JS macros I've added during the last year, so it seemed worthwhile to point those out.

**Comment by juj** [comment]
> FWIW (again, doesn't solve the library use-case, just for the record) yet another alternative would be to use the mentioned `EM_ASYNC_JS` from C++.

The EM_ASYNC_JS option would unfortunately not work for two reasons:

1) we do not consider EM_ASM or EM_JS (and hence by extension EM_ASYNC_JS) a production quality feature because of issue #9366 / #12778 .

2) The WebGPU library is being used with Closure, so EM_ASYNC_JS scheme would break with that in ASSERTIONS builds.

Though for curiosity, raised https://github.com/google/closure-compiler/issues/3900 to discuss if Closure would be able to accommodate for this kind of `__asyncjs__` naming scheme.

(it is uncommon, but still sometimes important to be able to enable assertions in release builds, e.g. to debug when some release mode optimization mysteriously fails)

**Comment by RReverser** [comment]
> 1. we do not consider EM_ASM or EM_JS (and hence by extension EM_ASYNC_JS) a production quality feature because of issue [Why are JavaScript inline snippets embedded in wasm data? #9366](https://github.com/emscripten-core/emscripten/issues/9366) / [EM_JS emits an odd data section to compiled code #12778](https://github.com/emscripten-core/emscripten/issues/12778) .

Ouch, that does look bad, I thought we stripped those. @sbc100 you mentioned in one of the issues you had a promising fix?

> The WebGPU library is being used with Closure, so EM_ASYNC_JS scheme would break with that in ASSERTIONS builds.
(it is uncommon, but still sometimes important to be able to enable assertions in release builds, e.g. to debug when some release mode optimization mysteriously fails)

Hmm, but then not only `__asyncjs__*` functions should be affected, but any functions in the `ASYNCIFY_IMPORTS` list as the linked snippet used `var isAsyncifyImport = ASYNCIFY_IMPORTS.indexOf(x) >= 0` even before I added the 2nd part. How come other Asyncify functions aren't broken by Closure?

**Comment by RReverser** [comment]
> > 1. we do not consider EM_ASM or EM_JS (and hence by extension EM_ASYNC_JS) a production quality feature because of issue [Why are JavaScript inline snippets embedded in wasm data? #9366](https://github.com/emscripten-core/emscripten/issues/9366) / [EM_JS emits an odd data section to compiled code #12778](https://github.com/emscripten-core/emscripten/issues/12778) .
> 
> Ouch, that does look bad, I thought we stripped those. @sbc100 you mentioned in one of the issues you had a promising fix?

Ah wait, looks like the first issue was fixed in https://github.com/emscripten-core/emscripten/pull/13443 and 2nd was just a duplicate but forgotten to be closed. So today those data sections should be stripped as expected.

**Comment by juj** [comment]
Oh, thanks for digging up that update - closing the old issue.

**Comment by juj** [comment]
It looks like this was addressed in PR #19093, so closing. Big thanks @brendandahl for implementing this!

**Comment by RReverser** [comment]
Note that there are some issues with marking `async` functions for now. Until we figure out automatic wrapping, it's probably best to only rely on `__async` annotation. https://github.com/emscripten-core/emscripten/issues/20289
